tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/anndata.py

```diff
@@ -19,99 +23,244 @@
-            missing_from_var = ad.raw.var.index.difference(ad.var.index)
-            if len(missing_from_var) > 0:
-                raw_var = ad.raw.var.loc[missing_from_var].copy()
-                raw_var["feature_is_filtered"] = True
-                # TODO - these should be looked up in the ontology
-                raw_var["feature_name"] = "unknown"
-                raw_var["feature_reference"] = "unknown"
-                raw_var["feature_length"] = 0
-                var = pd.concat([ad.var, raw_var])
-            else:
-                var = ad.raw.var
+    # else confusion
+    raise IndexError("Unsupported indexing types")
 
-        else:
-            X = ad.X
-            var = ad.var
 
-        if need_X and isinstance(X, (sparse.csr_matrix, sparse.csc_matrix)) and not X.has_canonical_format:
-            logging.warning(f"H5AD with non-canonical X matrix at {path}")
+def _normed_index(idx: Index) -> tuple[Index1D, Index1D]:
+    if not isinstance(idx, tuple):
+        return idx, slice(None)
+    elif len(idx) == 1:
+        return idx[0], slice(None)
+    elif len(idx) == 2:
+        return idx
+    else:
+        raise IndexError("Indexing supported on two dimensions only")
+
+
+class AnnDataProxy:
+    """
+    Recommend using `open_anndata2()` rather than instantiating this class directly.
+
+    AnnData-like proxy for the version 0.1.0 AnnData H5PY file encoding (aka H5AD).
+    Used in lieu of the AnnData class to reduce memory overhead. Semantics very similar
+    to anndata.read_h5ad(backed="r"), but with the following optimizations:
+
+    * opening and indexing does not materialize X (including repeated indexing,
+      which is an operation where AnnData always materializes X)
+    * opening only materializes obs/var, never obsm/varm/obsp/varp (which in many
+      datasets is very large and memory-intensive)
+    * only one copy of the base obs/var data is maintained. All views are defined
+      only by an index (either a Python range, or a numpy point index array)
+
+    This effectively:
+    * removes the overhead of reading unused slots such as obsm/varm/obsp/varp
+    * removes the overhead when you need to index a view (where AnnData materializes)
+
+    In the future, if we need additional slots accessible, we may need to add sub-proxies
+    for uns, obs*, var*, etc. But at the moment, the Builder has no use for these so they
+    are omitted.
+    """
+
+    _obs: pd.DataFrame
+    _var: pd.DataFrame
+    _X: h5py.Dataset | CSRDataset | CSCDataset
+
+    def __init__(
+        self,
+        filename: str | PathLike[str],
+        *,
+        view_of: Self | None = None,
+        obs_idx: slice | npt.NDArray[np.int64] | None = None,
+        var_idx: slice | npt.NDArray[np.int64] | None = None,
+        obs_column_names: Optional[Tuple[str, ...]] = None,
+        var_column_names: Optional[Tuple[str, ...]] = None,
+    ):
+        self.filename = filename
+        self.is_view = view_of is not None
```

Comment:
not used; keep anyway?

Reply:
I think there is a use for it in some scenarios, but not in the builder - will delete in the name of simplicity